chainer/function.py

```diff
@@ -55,111 +55,129 @@ def force_backprop_mode():
-       >>> x = chainer.Variable(numpy.zeros(10))
-       >>> f = F.Identity()
-       >>> y = f(x)
+    _function = None
+    _weak_function = None
 
-       computes a new variable ``y`` and creates backward references. Actually,
-       backward references are set as per the following diagram::
+    def __init__(self, function):
+        super(FunctionAdapter, self).__init__()
+        self._weak_function = weakref.ref(function)
+        function._owned_node = self
 
-           x.node <--- f <--- y.node
+    @property
+    def function(self):
+        """The :class:`Function` object that this adapter is wrapping."""
+        func = self._function
+        if func is not None:
+            return func
 
-       If an application of another function ``g`` occurs as
+        weak_func = self._weak_function
+        return weak_func and weak_func()
 
-       >>> g = F.Identity()
-       >>> z = g(x)
+    @property
+    def label(self):
+        return self._function.label
 
-       then the graph grows with a branch::
+    @property
+    def _impl_name(self):
+        return self._function.__class__.__name__
 
-                    |--- f <--- y.node
-           x.node <-+
-                    |--- g <--- z.node
+    def check_type_forward(self, in_types):
+        self._function.check_type_forward(in_types)
 
-       Note that the branching is correctly managed on backward computation,
-       i.e. the gradients from ``f`` and ``g`` are accumulated to the gradient
-       of ``x``.
+    def forward(self, inputs):
+        # Retain all inputs by default in old-style functions.
+        self.retain_inputs(six.moves.range(len(inputs)))
+        return self._function.forward(inputs)
+
+    def backward(self, target_input_indexes, grad_outputs):
+        in_data = tuple([input.data for input in self.inputs])
+        grad_out_data = tuple([None if grad is None else grad.data
+                               for grad in grad_outputs])
 
-    Every function implementation should provide :meth:`forward_cpu`,
-    :meth:`forward_gpu`, :meth:`backward_cpu` and :meth:`backward_gpu`.
-    Alternatively, one can provide :meth:`forward` and :meth:`backward` instead
-    of separate methods. Backward methods have default implementations that
-    just return ``None``, which indicates that the function is non-
-    differentiable.
+        gxs = self._function.backward(in_data, grad_out_data)
+        for x, gx in six.moves.zip(self.inputs, gxs):
+            variable._check_grad_type(self, x, gx)
 
-    For functions that do not need a part of inputs in backward computation,
-    there is a way to possibly reduce the memory consumption by quickly
-    releasing the input arrays after the forward propagation. This is done by
-    calling :meth:`retain_inputs` from inside of :meth:`forward` (including
-    :meth:`forward_cpu` and :meth:`forward_gpu`). See the documentation of
-    :meth:`retain_inputs` for details.
+        return tuple([
```

Comment:
In my understand when a user uses old-style Function, this gradient Variable does not have creator link. Is it right?